contracts/cw3-dao/src/state.rs

```diff
@@ -128,27 +121,112 @@ impl Proposal {
                 if self.expires.is_expired(block) {
                     // If expired, we compare vote_count against the total number of votes (minus abstain).
                     let opinions = self.votes.total() - self.votes.abstain;
-                    vote_count >= votes_needed(opinions, threshold)
+                    self.votes.yes >= votes_needed(opinions, threshold)
```

Comment:
Re: "failure on abstain" - if `opinions` is zero here `votes_needed(0, N)` will return `0` for all values of `N`. This means that if all the votes are abstain and `opinions` is zero a proposal will be considered passed even if `self.votes.yes` is zero as `0 >= 0`.